---
title: "Import scRNA Data"
author: "Jianyu Liu"
date: "November 28, 2018"
output:
  pdf_document: default
  html_document: default
---

```{r, echo=F}
library(HUG)
```


This document describes how I import and preprocess scRNA-seq data and others for the paper.

# scRNA-seq Data
## Velmeshev Data
Paper:  Single-cell genomics identifies cell typeâ€“specific molecular changes in autism
Source: <https://cells.ucsc.edu/autism/rawMatrix.zip>
Raw data file: ```barcodes.tsv``` saves barcode information, ```genes.tsv``` saves gene information, and ```matrix.mtx``` saves the count data in MatrixMarket format
Technique: function ```read10xCounts``` from R package ```DropletUtils```

```{r Velmeshev}
data.dir = "Data/autism"
sce      = read10xCounts(file.path(data.dir, "rawMatrix"), col.names=TRUE)
cell_info = fread(file.path(data.dir, "meta.tsv"))
colData(sce) = cbind(colData(sce), cell_info)
is.mito = which(rowData(sce)$chromosome_name == "MT")
is.ribo = which(rowData(sce)$hgnc_symbol %in% ribo$'Approved Symbol')
sce = calculateQCMetrics(sce, feature_controls=list(Mt=is.mito, Ri=is.ribo))
names(rowData(sce))[6] = "strand_n"
sce = sce[which(rowData(sce)$n_cells_by_counts >= 0.01*ncol(sce)),]
u.regions  = unique(colData(sce)$region)
u.clusters = unique(colData(sce)$cluster)

data.dir = "Data/autism"
for(r1 in u.regions){
  cond1 = colData(sce)$region == r1
  for(cls1 in u.clusters){
    cond2 = colData(sce)$cluster == cls1
    ct1   = counts(sce[,which(cond1 & cond2)])
    cls2  = gsub("/", "_", cls1)
    fnm   = sprintf("%s_%s.rds", r1, cls2)
    saveRDS(ct1, file=file.path(data.dir, "ct_mtx", fnm))
  }
}
names = c("L2_3","IN-VIP","IN-SV2C","IN-SST","IN-PV")
name.post = "PFC_"
data.people = levels(factor(read.table("Data/autism/meta.tsv",sep = '\t',header = TRUE)$individual))
data.save = vector("list",length(data.people)+1)
gene.num = nrow(as.matrix(readRDS(paste0("Data/autism/ct_mtx/",name.post,names[1],".rds"))))
names(data.save) = c(data.people,"all")

print(paste0("processing ",names[1]))
data.all = t(as.matrix(readRDS(paste0("Data/autism/ct_mtx/",name.post,names[1],".rds"))))

for (i in 2:length(names)){
  print(paste0("processing ",names[i]))
  data.now = t(as.matrix(readRDS(paste0("Data/autism/ct_mtx/",name.post,names[i],".rds"))))
  data.all = rbind(data.all,data.now)
  print(paste0(i,"/",length(names)," dataset complete"))
}
save(data.all,file = paste0("Data/autism/ct_mtx/","Velmeshev_all.rData"))


```


## Gierahn Data
Paper: Seq-Well: A portable, low-cost platform for single-cell RNA-seq of low-input samples
Source: <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE92495>
Raw data file: GSE92495_RAW.tar/GSM2486331_HEK_SingleCells.txt
Size: 1,453 cells * 24,187 genes
Technique: Seq-Well based on 

```{r gierahn}
dat.gierahn = read.table('Data/GSM2486331_HEK_SingleCells.txt')
dat.gierahn = Matrix(t(dat.gierahn))
ix.gierahn = as.factor(sapply(strsplit(rownames(dat.gierahn), '_'), `[`, 1))
mean(dat.gierahn == 0)
save(dat.gierahn, ix.gierahn, file='Data/Gierahn2017_raw.rData')
```


# Benchmark Graph
```{r benchmark}
conns = read.table('Data/PathwayCommons9.All.hgnc.sif', header=F, as.is=T)
genes = unique(c(conns$V1, conns$V3))
rel = unique(conns$V2)
p = length(genes)

conns$V1 = as.integer(factor(conns$V1, levels=genes))
conns$V3 = as.integer(factor(conns$V3, levels=genes))

rels.gene = foreach(k=rel) %do% with(conns[conns$V2==k,], 
                                     sparseMatrix(i=V1, j=V3, x=T, dims=c(p, p), dimnames=list(genes, genes)))
conns.gene = lapply(rels.gene, function(x) x | t(x))
conn.gene = Reduce('|', conns.gene)
save(conn.gene, file='Data/PathwayCommons9.rData')
```

# Gene Lists
We create a list of genes that are expressed in >= 30% of the cells in both datasets. Moreover, we remove RPS- and RPL-genes because they correspond to a misleading dense sub-graph in the benchmark.

```{r gene1960}
genes = Reduce(intersect, lapply(list(dat.gierahn, dat.tirosh, conn.gene), colnames))
nz.gierahn = colMeans(dat.gierahn[,genes] >= .5)
nz.tirosh = colMeans(dat.tirosh[,genes] >= .5)

genes.tmp = genes[startsWith(genes, 'RPL') | startsWith(genes, 'RPS')]
image.matrix(conn.gene[genes.tmp, genes.tmp])
genes.now = genes[(nz.gierahn >= .3) & (nz.tirosh >= .3) &
                      !(startsWith(genes, 'RPL') | startsWith(genes, 'RPS'))]
save(genes.now, file='genes_1960.rData')
```

We also create a list of RB-related genes based on the C6 database. 
```{r geneRB}
name.pathway = c("RB_DN.V1_DN", "RB_DN.V1_UP", "RB_P107_DN.V1_DN", "RB_P107_DN.V1_UP", 
                 "RB_P130_DN.V1_DN", "RB_P130_DN.V1_UP")
pathways = list()
for (i in 1:189) {
    txt.i = scan("Data/c6.all.v5.2.symbols.gmt.txt", quiet = TRUE, 
                 skip = i-1, nlines = 1, what = character())
    pathways[[txt.i[1]]] = txt.i[-(1:2)]
}
paths.RB = pathways[name.pathway]
genes.RB = Reduce(union, gene.lists0) # 2354 genes
save(genes.RB, paths.RB, file='genes_RB.rData')
```
